File: Kotlin/four/src/main/kotlin/com/nlw/four/domain/service/CreateUserService.kt
package com.nlw.four.domain.service

import com.nlw.four.domain.exception.BusinessException
import com.nlw.four.domain.model.User
import com.nlw.four.domain.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class CreateUserService(
    @Autowired private val userRepository: UserRepository
) {
    fun create(user: User): User {
        val userExists = userRepository.findByEmail(user.email!!)

        if (userExists != null && !userExists.equals(user)) {
            throw BusinessException("User already exists")
        }

        return userRepository.saveAndFlush(user)
    }
}

File: Node/src/services/CreateUserService.ts
import { getRepository } from "typeorm";
import { User } from "../models";

class CreateUserService {
  async create(email: string, name: string): Promise<User> {
    const userRepository = getRepository(User);

    const userAlreadyExists = await userRepository.findOne({
      email
    });

    if (userAlreadyExists) {
      throw new Error('User already exists');
    }

    const user = userRepository.create({ name, email });

    await userRepository.save(user);

    return user;
  }
}

export { CreateUserService };
